### DISCUSSION TITLE
Testing: Capturing output

### DISCUSSION THREAD

**Discussion by owllex** [evennia/evennia#2499] [Community Contribs & Snippets]
Hey everyone,

I know, testing is everyone's favorite topic, but it's important to keep your project bug-free if you're in it for the long-haul. I've recently been working on unit and integration testing for my own project and wanted to come up with a more fluent and pythonic way of asserting against expected output for one or more characters/sessions.

What I wanted to be able to do was to run some code (for example, a command, or something in a script that generated text output) and then assert that one or more sessions observing that output saw the right thing.

Originally I thought it would be sufficient just to mock DefaultObject.msg, but doing so actually skips a few very important things: pre- and post-msg hooks. In my codebase, some important stuff happens here like certain classes of text substitution or formatting that I didn't want to mock out. Asserting against output that didn't pass through those hooks was not terribly fluent and didn't make for a good integration test as it didn't actually reflect what a user was going to see.

Mocking the session itself seemed like a better option, but there were still problems. If one method call generated output to multiple sessions (say, something that triggered a `msg_contents`), you'd need to sequentially assert for each receiver. If one of those assertions fails, the error message only gives you a small piece of the overall context. And moreover, if you wanted to assert against multiple sessions of output over multiple stages, you'd need to manually reset all relevant mocks between each call. Yes, it's doable, but it's more boilerplate that distracts from what the test is actually doing. There should be a utility to do that.

So I wrote one. Below is a context manager that patches `data_out()` calls for attached sessions for all provided objects, and then produces a single dict of all relevant results that you can assert against in one shot. Since it's a context manager, it cleans itself up when it's done, so you can use multiple instances of it in your test to isolate outputs from different calls to methods under test cleanly and with minimal boilderplate.

```python
class capture_output(object):
    """
    Context manager used to patch session output sent to objects.
    This can be used to assert expectations against text output to one or
    more receiver objects supplied in the constructor with a single
    assert statement, rather than asserting against multiple mocks in
    sequence.
    Usage:
        with capture_output(obj1, obj2, obj3) as captured:
            # ... Perform calls that generate output.
            self.assertEqual(
                captured.output(),
                {obj1: "Obj1 text", obj2: "Obj2 text", obj3: None}
            )
    """

    def __init__(self, *args):
        """
        Initialize this manager with one or more objects to capture from.
        Args:
            *args: One or more positional args, each of which should be an
                object with an attached session. Output sent to the first
                such session for each object will be captured within the
                scope of this context manager.
        """
        self.obj = args

    def __enter__(self):
        self.patches = {}
        # Upon entry, mock output sessions attached to each object.
        for obj in self.obj:
            sessions = obj.sessions.all() if obj.sessions else None
            if sessions and (session := sessions[0]):
                backup = session.data_out
                mock = Mock()
                session.data_out = mock
                self.patches[obj] = (backup, mock)
        return self

    def __exit__(self, *args):
        for obj, (backup, mock) in self.patches.items():
            sessions = obj.sessions.all() if obj.sessions else None
            if sessions and (session := sessions[0]) and session.data_out == mock:
                session.data_out = backup
        return False

    def output(self, noansi=True):
        """
        Returns captured output.
        Args:
            noansi (bool): True to strip color tags from output. This makes
                assertions more readable if color information does not matter.
        Returns:
            (dict): A dictionary with a key for each object specified when this
                context manager was entered. Each key maps to a string which
                contains all of the output received since the context manager
                was opened, separated by newlines if more than one was present.
                If no output was received for a given object, its value is
                None.
        """
        result = {}
        for obj, (_, mock) in self.patches.items():
            if calls := mock.mock_calls:
                messages = [
                    kwargs.get("text", "") for name, args, kwargs in calls if kwargs
                ]
                messages = [
                    str(text[0]) if isinstance(text, tuple) else str(text)
                    for text in messages
                    if text
                ]
                text = "\n".join(messages)
                text = ansi.parse_ansi(text, strip_ansi=noansi).strip()
                result[obj] = text
            else:
                result[obj] = None

        if len(result) == 1:
            result = next(iter(result.values()))
        return result
```

You can use this on its own, but you can also combine it with your own `CommandTest` subclass to enable fluent assertions for command output as well. For example, something like the following:

```python
    # Inside CommandTest.call
        with capture_output(*receivers) as capture:
            try:
                if command.at_pre_cmd():
                    return capture.output(noansi=noansi)
                command.parse()
                command.func()
                command.at_post_cmd()
            except InterruptCommand:
                pass
            return capture.output(noansi=noansi)
```

And then you can write assertions like this:
```python
    # Inside my CmdEmote tests
        self.assertEqual(
            self.call("emote looks at /ari:'s outfit.", rel, ari, other),
            {
                rel: "Rel looks at Ari's outfit.",
                ari: "Rel looks at your outfit.",
                other: "A male elf looks at a female human's outfit.",
            },
        )
```

Hopefully this is useful to someone. It's saved me tons and tons of time and boilerplate when writing tests.

**Comment by Griatch** [comment]
This looks really interesting!
